### ISSUE TITLE
systemd seems to be failing to start systemd-resolved inside an unprivileged container

### ISSUE THREAD

**Issue by evverx** [systemd/systemd#9914]
Since 1beab8b0d0ff2d7d1 was merged, `systemd` has been failing to start `systemd-resolved` inside unprivileged containers:
```sh
Aug 20 21:07:07 systemd-testsuite systemd[23]: Not remounting /run/systemd/unit-root/var/tmp, because blacklisted by /run/systemd/unit-root/var/tmp, called for /run/systemd/unit-root
Aug 20 21:07:07 systemd-testsuite systemd[23]: Failed to re-mount '/run/systemd/unit-root/dev' read-only: Operation not permitted
Aug 20 21:07:07 systemd-testsuite systemd[23]: systemd-resolved.service: Failed to set up mount namespacing: Operation not permitted
Aug 20 21:07:07 systemd-testsuite systemd[23]: systemd-resolved.service: Failed at step NAMESPACE spawning /usr/lib/systemd/systemd-resolved: Operation not permitted
Aug 20 21:07:07 systemd-testsuite systemd[1]: systemd-resolved.service: Child 23 belongs to systemd-resolved.service.
Aug 20 21:07:07 systemd-testsuite systemd[1]: systemd-resolved.service: Main process exited, code=exited, status=226/NAMESPACE
Aug 20 21:07:07 systemd-testsuite systemd[1]: systemd-resolved.service: Failed with result 'exit-code'.
Aug 20 21:07:07 systemd-testsuite systemd[1]: systemd-resolved.service: Changed start -> failed
Aug 20 21:07:07 systemd-testsuite systemd[1]: systemd-resolved.service: Job systemd-resolved.service/start finished, result=failed
Aug 20 21:07:07 systemd-testsuite systemd[1]: Failed to start Network Name Resolution.
Aug 20 21:07:07 systemd-testsuite systemd[1]: systemd-resolved.service: Unit entered failed state.
```
The container was launched with `systemd-nspawn -U`.

**Comment by sourcejedi** [comment]
I looked at the source code, but I can't work out how this happens.  Based on the exact message (no "and its submounts"), it seems the failure must be this line:

https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L1047

        } else if (m->mode == PRIVATE_DEV) {
                /* Superblock can be readonly but the submounts can't */
                if (mount(NULL, mount_entry_path(m), NULL, MS_REMOUNT|DEV_MOUNT_OPTIONS|MS_RDONLY, NULL) < 0)

1) I can't see how this can fail. The `dev` mount was just mounted by `mount_private_dev()` (i.e. in the current user namespace).  So we should be allowed to remount it.

2) I can't even see how this line is reached. The PRIVATE_DEV entry is added without setting `read_only`. And I can't see anywhere that systemd adds a read only MountEntry for "/dev".

**Comment by evverx** [comment]
One does not simply remount `tmpfs` read-only in user namespaces :-)
```sh
$ unshare -U -r -m -p --fork --mount-proc
# mkdir TMP
# mount -t tmpfs tmpfs TMP/
# mount -o remount,ro TMP/
mount: ./TMP: permission denied.
```
I figure `MS_BIND` should be enough to make that `mount` work in unprivileged containers.

**Comment by evverx** [comment]
By the way, you're correct about the code triggering the issue. I've just put `MS_BIND` right after `MS_REMOUNT` there and the failure is gone, but It doesn't seem to be the only place needs fixing, though. I'll try to look into this later this week.

**Comment by sourcejedi** [comment]
> One does not simply remount tmpfs read-only in user namespaces :-)

Actually it looks like it was quite simple :-P. https://github.com/torvalds/linux/commit/bc6155d13260

> I've just put MS_BIND right after MS_REMOUNT there and the failure is gone

Sounds good.  It's basically my fault :-).  I pointed out we could remount read-only without `MS_BIND` when reviewing the PR that added `TemporaryFilesystem=`, and poettering suggested to change `PrivateDevices=` at the same time.

I think it's safe to change back, and I don't expect anyone will notice a difference in behaviour.  

It just surprised me to realize that `TemporaryFilesystem=/tmp:size=10M,ro,nosuid` would not apply `ro` to the superblock (underlying filesystem), like `mount -osize=10M,ro,nosuid` does.  Maybe a comment could note the kernel version (v4.18), that lets you remount without `MS_BIND` inside a user namespace.  For the benefit of any other reader who is as pedantic as I was.

**Comment by evverx** [comment]
It seems you're much more familiar with this part of the code. Could you send a PR? Meanwhile, I'll try to figure out why the test didn't fail as expected on `s390x` in https://github.com/systemd/systemd/pull/9957.

> For the benefit of any other reader who is as pedantic as I was.

I seriously doubt pedantic readers trust what is written in comments :-) It's not to say that they shouldn't be added.

**Comment by evverx** [comment]
I launched a VM with Ubuntu Bionic (4.13.0-25). Apparently, the commit you pointed out has already been backported, which probably shouldn't be surprising given who the authors and reviewers are.

**Comment by evverx** [comment]
Just visited https://unix.stackexchange.com/questions/465202/inside-a-user-namespace-why-am-i-not-allowed-to-remount-a-filesystem-i-have-mou. Couldn't resist making a loop by posting the link here :-)

**Comment by yuwata** [comment]
You may already know but I note that `MS_BIND` is used in `bind_remount_recursive_with_mountinfo()`.

**Comment by xnox** [comment]
Hi, i was away for a while. Is this just a kernel issue then which will be sorted out when we get new kernels everywhere on the cloud images?

**Comment by sourcejedi** [comment]
@xnox I think we want to fix it (i.e. go back to remounting with MS_BIND), because we don't list kernel v4.18 as a requirement yet :-).

I'll have a go at this.

@yuwata thanks for pointing that out.  `PRIVATE_DEV` explicitly doesn't want a recursive remount.

I started looking at reducing some duplication... which brought in some more "fun" side-issues to think about.  ([Temporary link, it looks like they're going to delete this](https://stackoverflow.com/questions/52057444/can-root-reliably-determine-the-per-mount-bind-mount-flags-e-g-for-a-fuse-f?noredirect=1#comment91065334_52057444)).

**Comment by evverx** [comment]
@xnox http://kernel.ubuntu.com/git/ubuntu/ubuntu-bionic.git/commit/fs/namespace.c?id=1db30ec055c3d9540bd08923d79173a63b06ea62 landed in 2016, so, in a way, it's already been sorted out.

@sourcejedi I'm not sure what duplication you're talking about :-) Everything is fine as long as it works :-) Seriously though, regarding `statvfs`, the comment in the snippet below seems to indicate that it isn't always enough for `NFS` at least
https://github.com/systemd/systemd/blob/e094eb11888d984f1e9139675a5780a49c61d5ed/src/basic/stat-util.c#L121-L138

**Comment by sourcejedi** [comment]
I've written up my "fun" side investigation as a separate issue.

The regression fix is in #9977.

**Comment by evverx** [comment]
@sourcejedi thank you! That's a lot to process :-) I'll take a look on Monday. I don't have my laptop close at hand right now.

**Comment by yayoi-k** [comment]
I'm presently seeing the issue described here when containers are launched with systemd-nspawn -U.  My containers are Ubuntu 22.04 based.

Logs from inside the container:
root@container00:~# journalctl -f -u systemd-resolved.service
May 11 15:01:57 container00 systemd-resolved[86]: Failed to listen on UDP socket 127.0.0.53:53: Permission denied
May 11 15:01:57 container00 systemd-resolved[86]: Failed to start manager: Permission denied
May 11 15:01:57 container00 systemd[1]: systemd-resolved.service: Main process exited, code=exited, status=1/FAILURE
May 11 15:01:57 container00 systemd[1]: systemd-resolved.service: Failed with result 'exit-code'.
May 11 15:01:57 container00 systemd[1]: Failed to start Network Name Resolution.
May 11 15:01:57 container00 systemd[1]: systemd-resolved.service: Scheduled restart job, restart counter is at 5.
May 11 15:01:57 container00 systemd[1]: Stopped Network Name Resolution.
May 11 15:01:57 container00 systemd[1]: systemd-resolved.service: Start request repeated too quickly.
May 11 15:01:57 container00 systemd[1]: systemd-resolved.service: Failed with result 'exit-code'.
May 11 15:01:57 container00 systemd[1]: Failed to start Network Name Resolution.

nspawn output when launching the container:
<img width="569" alt="Unknown-2" src="https://github.com/systemd/systemd/assets/2354113/8e97c0c8-e6a3-448a-b9fd-08acc544b0e2">

**Comment by xnox** [comment]
@LogExE this issue from 2018 is unlikely to be related. Issues in 22.04 are not appropriate to report to systemd upstream. Instead please use `ubuntu-bug systemd` to report issue to Ubuntu first.
